src/operator/contrib/ngraph.cc

```diff
@@ -0,0 +1,266 @@
+}
+
+void NgraphSubgraphOpForward(const OpStatePtr& state_ptr, const OpContext& ctx,
+                             const std::vector<NDArray>& inputs,
+                             const std::vector<OpReqType>& req,
+                             const std::vector<NDArray>& outputs) {
+  NgraphSubgraphOperator& op = state_ptr.get_state<NgraphSubgraphOperator>();
+  op.Forward(ctx, inputs, req, outputs);
+}
+
+void NgraphSubgraphOpBackward(const OpStatePtr& state_ptr, const OpContext& ctx,
+                              const std::vector<NDArray>& inputs,
+                              const std::vector<OpReqType>& req,
+                              const std::vector<NDArray>& outputs) {
+  NgraphSubgraphOperator& op = state_ptr.get_state<NgraphSubgraphOperator>();
+  op.Backward(ctx, inputs, req, outputs);
+}
+
+std::vector<nnvm::NodeEntry> NgraphSubgraphGradient(
+    const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
+  auto graph = get_ngraph(n->attrs);
+  const bool zero_grad = check_zero_grad(graph);
+  graph->zero_grad = zero_grad;
+  auto is_loss = graph->is_loss;
+  auto p = nnvm::Node::Create();
+  p->attrs.op = nnvm::Op::Get("_backward_ngraph_subgraph_op");
+  p->attrs.parsed = n->attrs.parsed;
+  if (std::find(begin(is_loss), end(is_loss), true) == end(is_loss) &&
+      zero_grad && graph->num_outputs_ == 1) {
+    return mxnet::op::MakeZeroGradNodes(n, ograds);
+  }
+  p->attrs.name = n->attrs.name + "_backward";
+  p->attrs.dict = n->attrs.dict;
+  p->control_deps.emplace_back(n);
+  if (p->op()->attr_parser != nullptr) {
+    p->op()->attr_parser(&(p->attrs));
+  }
+  if (!zero_grad) {
+    for (size_t i = 0; i < ograds.size(); ++i) {
+      if (!is_loss[i]) {
+        p->inputs.push_back(ograds[i]);
+      }
+    }
+  }
+  p->inputs.insert(p->inputs.end(), n->inputs.begin(), n->inputs.end());
+  std::vector<nnvm::NodeEntry> ret;
+  for (unsigned i = 0; i < p->num_outputs(); ++i) {
+    ret.emplace_back(nnvm::NodeEntry{p, i, 0});
+  }
+  return ret;
+}
+
+NNVM_REGISTER_OP(_ngraph_subgraph_op)
+    .describe(R"code(_ngraph_subgraph_op)code" ADD_FILELINE)
+    .set_num_inputs([](const NodeAttrs& attrs) {
+      auto graph = get_ngraph(attrs);
+      return graph->inputs_.size();
+    })
+    .set_num_outputs([](const NodeAttrs& attrs) {
+      auto graph = get_ngraph(attrs);
+      return graph->outputs_.size();
+    })
+    .set_attr<nnvm::FListInputNames>("FListInputNames",
+                                     [](const nnvm::NodeAttrs& attrs) {
+                                       auto graph = get_ngraph(attrs);
+                                       std::vector<std::string> input_names;
+
+                                       for (auto n : graph->inputs_) {
+                                         input_names.emplace_back(n->name_);
+                                       }
+                                       return input_names;
```

Comment:
can you use the helper functions here for getting input/output names, infer attributes, and so on?

Reply:
Helper functions instead of lambdas?

Reply:
Helper functions instead of lambdas? -> Yes

Reply:
Can do. Thanks.